### PR TITLE
BC-3997 - Board Error Handling

### PR DESCRIPTION
**Pull request by muratmerdoglu-dp**
# Short Description
As a user, I want the board to work as expected and want to get meaningful error messages when there is an error.

## Links to Ticket and related Pull-Requests
https://ticketsystem.dbildungscloud.de/browse/BC-3997

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: src/components/feature-board/state/BoardState.composable.ts [hunk 1]
```diff
@@ -105,10 +132,16 @@ export const useBoardState = (id: string) => {
 	const fetchBoard = async (id: string): Promise<void> => {
 		isLoading.value = true;
 		const boardsApi = BoardApiFactory(undefined, "/v3", $axios);
-		board.value = {
-			...(await boardsApi.boardControllerGetBoardSkeleton(id)).data,
-			id,
-		};
+
+		const response = (await boardsApi.boardControllerGetBoardSkeleton(id)).data;
```
Comment (uidp) [inline src/components/feature-board/state/BoardState.composable.ts:136]: `response.data` is the created resource. If you want to access the response you should remove `.data`. The expected response status code for successfull operation is 200. So checking the status accordingly might make sense.

File: src/components/feature-board/state/BoardState.composable.ts [hunk 2]
```diff
@@ -105,10 +132,16 @@ export const useBoardState = (id: string) => {
...
+
+		if (!response?.id) {
```
Comment (uidp) [inline src/components/feature-board/state/BoardState.composable.ts:138]: Maybe better check for status code? Also `response.data` cannot be `undefined` (no question mark needed).

File: src/components/feature-board/state/BoardState.composable.ts [hunk 3]
```diff
@@ -154,13 +196,29 @@ export const useBoardState = (id: string) => {
...
 		await nextTick();
 		board.value.columns.splice(addedIndex, 0, element);
-		await moveColumnCall(payload.payload, board.value.id, addedIndex);
+		const response = await moveColumnCall(
+			payload.payload,
+			board.value.id,
+			addedIndex
+		);
+
+		if (isErrorCode(response?.status)) {
```
Comment (uidp) [inline src/components/feature-board/state/BoardState.composable.ts:205]: response is will never be `undefined`(no question mark needed)

File: src/components/feature-board/state/BoardState.composable.ts [hunk 4]
```diff
@@ -154,13 +196,29 @@ export const useBoardState = (id: string) => {
...
+		}
 	};
 
 	const updateColumnTitle = async (columnId: string, newTitle: string) => {
 		if (board.value === undefined) return;
 
-		await updateColumnTitleCall(columnId, newTitle);
+		const response = await updateColumnTitleCall(columnId, newTitle);
+
+		if (isErrorCode(response?.status)) {
```
Comment (uidp) [inline src/components/feature-board/state/BoardState.composable.ts:216]: response is will never be undefined(no question mark needed)

File: src/components/feature-board/state/BoardState.composable.ts [hunk 5]
```diff
@@ -137,7 +170,16 @@ export const useBoardState = (id: string) => {
 			if (card) {
 				addCard(card, targetColumnId, addedIndex);
 			}
-			await moveCardCall(payload.cardId, targetColumnId, addedIndex);
+			const response = await moveCardCall(
+				payload.cardId,
+				targetColumnId,
+				addedIndex
+			);
+			if (isErrorCode(response?.status)) {
```
Comment (uidp) [inline src/components/feature-board/state/BoardState.composable.ts:178]: response is will never be undefined(no question mark needed)

File: src/components/feature-board/state/BoardState.composable.ts [hunk 6]
```diff
@@ -51,31 +62,47 @@ export const useBoardState = (id: string) => {
...
+		};
+		await moveCard(moveCardPayload);
 	};
 
 	const deleteCard = async (id: string) => {
-		await deleteCardCall(id);
+		if (board.value === undefined) return;
+		const response = await deleteCardCall(id);
+
+		if (isErrorCode(response?.status)) {
```
Comment (uidp) [inline src/components/feature-board/state/BoardState.composable.ts:83]: The return type of `deleteCardCall` is `AxiosPromise<void>`, so `response` will never be `undefined` (no question mark needed)

File: src/components/feature-board/state/BoardState.composable.ts [hunk 7]
```diff
@@ -51,31 +62,47 @@ export const useBoardState = (id: string) => {
...
-
-		if (newColumn) {
-			const moveCardPayload: CardMove = {
-				addedIndex: 0,
-				removedIndex: null,
-				columnId: newColumn.id,
-				payload: { cardId: movingCardId, height: 100 },
-			};
-			await moveCard(moveCardPayload);
+		if (!newColumn?.id) {
```
Comment (uidp) [inline src/components/feature-board/state/BoardState.composable.ts:65]: Maybe it's better to check for the expected status code (see below).

File: src/components/feature-board/state/CardState.composable.ts [hunk 8]
```diff
@@ -16,19 +17,21 @@ export type AddCardElement = (
...
 		updateCardHeightCall,
 		updateCardTitle,
 	} = useBoardApi();
+	const { isErrorCode, showFailure, generateErrorText } = useBoardNotifier();
 
 	const fetchCard = async (id: string): Promise<void> => {
 		try {
 			cardState.card = await fetchCardFromApi(id);
 		} catch (error) {
+			const errorText = generateErrorText("read", "boardCard");
```
Comment (uidp) [inline src/components/feature-board/state/CardState.composable.ts:33]: What kind of exception is handled here? And why are we using exeptions for error handlich is the error handled by exceptions whereas we are using the api return value in the other places?

File: src/components/feature-board/state/CardState.composable.ts [hunk 9]
```diff
@@ -38,16 +41,21 @@ export const useCardState = (id: BoardCard["id"]) => {
 		if (cardState.card === undefined) {
 			return;
 		}
-		await updateCardTitle(cardState.card.id, newTitle);
+		const response = await updateCardTitle(cardState.card.id, newTitle);
+		if (isErrorCode(response?.status)) {
```
Comment (uidp) [inline src/components/feature-board/state/CardState.composable.ts:45]: `response` will never be `undefined`.

File: src/components/feature-board/state/CardState.composable.ts [hunk 10]
```diff
@@ -38,16 +41,21 @@ export const useCardState = (id: BoardCard["id"]) => {
...
 
 	const deleteCard = async () => {
-		if (cardState.card === undefined) {
-			return;
-		}
+		if (cardState.card === undefined) return;
 
-		await deleteCardCall(cardState.card.id);
+		const response = await deleteCardCall(cardState.card.id);
+		if (isErrorCode(response?.status)) {
```
Comment (uidp) [inline src/components/feature-board/state/CardState.composable.ts:56]: `response` will never be `undefined`.

File: src/components/feature-board/state/CardState.composable.ts [hunk 11]
```diff
@@ -66,20 +78,29 @@ export const useCardState = (id: BoardCard["id"]) => {
 			return;
 		}
 		const result = await createElement(cardState.card.id, { type });
-
+		if (!result.id) {
```
Comment (uidp) [inline src/components/feature-board/state/CardState.composable.ts:81]: Maybe better to check by repose status.

**Review by uidp — CHANGES_REQUESTED** [review]
I think we should think about how we can handle unsuccessful operations on the API more consistently. As far as I can see we are using different approaches:
- check the response status code
- check the response itself (resource) or its `id`
- use exceptions
Using the response value as an indicator for an unsuccessful operation doesn't look like a good approach IMHO.

File: src/components/feature-board/state/BoardState.composable.ts [hunk 12]
```diff
@@ -195,6 +253,13 @@ export const useBoardState = (id: string) => {
 		return columnIndex;
 	};
 
+	const showErrorAndReload = async (errorText: string | undefined) => {
+		if (board.value === undefined) return;
+
+		showFailure(errorText);
+		await fetchBoard(board?.value.id);
```
[the remark this answers is not in the document]
Comment (muratmerdoglu-dp) [reply src/components/feature-board/state/BoardState.composable.ts:260]: Considering card moving, the move card operation is triggered by the user for example, the card is in the new position in board object but there is an error on backend call. In this case we need to fetch board data from the backend for consistency. So I think we should keep it as it is.

File: src/components/feature-board/content-elements/FileContentElement.unit.ts [hunk 13]
```diff
@@ -1,27 +1,37 @@
...
+	};
+	const notifierModule = createModuleMocks(NotifierModule);
 	const getWrapper = (props: {
 		element: AnyContentElement;
 		isEditMode: boolean;
 	}) => {
 		document.body.setAttribute("data-app", "true");
 
-		const wrapper = shallowMount(FileContentElement, {
+		const wrapper = shallowMount(FileContentElement as MountOptions<Vue>, {
```
[the remark this answers is not in the document]
Comment (muratmerdoglu-dp) [reply src/components/feature-board/content-elements/FileContentElement.unit.ts:32]: My linter complains about if there is no casting.

**Review by uidp — APPROVED** [review]

**Review by uidp — APPROVED** [review]

**Pull request merged**